state-chain/runtime/src/chainflip/bitcoin_elections.rs

```diff
@@ -284,6 +297,120 @@ impl Hook<btc::BlockNumber, Vec<DepositChannelDetails<Runtime, BitcoinInstance>>
+	type BWProcessorTypes = BlockVaultWitnessingProcessorDefinition; // new block processor with Self::Blockdata
+}
+
+/// Associating the ES related types to the struct
+impl ElectoralSystemTypes for BitcoinVaultDepositWitnessingDefinition {
+	type ValidatorId = <Runtime as Chainflip>::ValidatorId;
+	type ElectoralUnsynchronisedState = BWState<BitcoinVaultDepositWitnessingDefinition>;
+	type ElectoralUnsynchronisedStateMapKey = ();
+	type ElectoralUnsynchronisedStateMapValue = ();
+	type ElectoralUnsynchronisedSettings = BWSettings;
+	type ElectoralSettings = ();
+	type ElectionIdentifierExtra = ();
+	type ElectionProperties = (BlockNumber, ElectionPropertiesVaultDeposit, u8);
+	type ElectionState = ();
+	type VoteStorage = vote_storage::bitmap::Bitmap<
+		ConstantIndex<(BlockNumber, ElectionPropertiesVaultDeposit, u8), BlockDataVaultDeposit>,
+	>;
+	type Consensus =
+		ConstantIndex<(BlockNumber, ElectionPropertiesVaultDeposit, u8), BlockDataVaultDeposit>;
+	type OnFinalizeContext = Vec<ChainProgress<BlockNumber>>;
+	type OnFinalizeReturn = Vec<()>;
+}
+
+/// Associating the state machine and consensus mechanism to the struct
+impl StateMachineES for BitcoinVaultDepositWitnessingDefinition {
+	// both context and return have to be vectors, these are the item types
+	type OnFinalizeContextItem = ChainProgress<BlockNumber>;
+	type OnFinalizeReturnItem = ();
+
+	// restating types since we have to prove that they have the correct bounds
+	type Consensus2 =
+		ConstantIndex<(BlockNumber, ElectionPropertiesVaultDeposit, u8), BlockDataVaultDeposit>;
+	type Vote2 =
+		ConstantIndex<(BlockNumber, ElectionPropertiesVaultDeposit, u8), BlockDataVaultDeposit>;
+	type VoteStorage2 = vote_storage::bitmap::Bitmap<
+		ConstantIndex<(BlockNumber, ElectionPropertiesVaultDeposit, u8), BlockDataVaultDeposit>,
+	>;
+
+	// the actual state machine and consensus mechanisms of this ES
+	type StateMachine = BWStateMachine<BitcoinVaultDepositWitnessingDefinition>;
+	type ConsensusMechanism =
+		BWConsensus<BlockDataVaultDeposit, BlockNumber, ElectionPropertiesVaultDeposit>;
+}
+
+/// Generating the state machine-based electoral system
+pub type BitcoinVaultDepositWitnessing =
+	StateMachineESInstance<BitcoinVaultDepositWitnessingDefinition>;
+
+#[derive(
+	Clone,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Debug,
+	Encode,
+	Decode,
+	TypeInfo,
+	MaxEncodedLen,
+	Serialize,
+	Deserialize,
+	Default,
+)]
+pub struct BitcoinVaultDepositWitnessingGenerator;
+
+use crate::chainflip::address_derivation::btc::derive_current_and_previous_epoch_private_btc_vaults;
+
+impl Hook<btc::BlockNumber, ElectionPropertiesVaultDeposit>
+	for BitcoinVaultDepositWitnessingGenerator
+{
+	fn run(&mut self, _block_witness_root: BlockNumber) -> ElectionPropertiesVaultDeposit {
+		let properties = pallet_cf_swapping::BrokerPrivateBtcChannels::<Runtime>::iter()
+			.flat_map(|(broker_id, channel_id)| {
+				derive_current_and_previous_epoch_private_btc_vaults(channel_id)
+					.unwrap()
```

Comment:
Let's not unwrap here, we can instead use if let Err and log_or_panic just for defensive purposes (ik it's basically impossible it's going to happen)